full-stack-intro/client/src/components/Todos.tsx

```diff
@@ -0,0 +1,156 @@
+import { useEffect, useState } from 'react';
+import { PageTitle } from './PageTitle';
+import { TodoList } from './TodoList';
+import { TodoForm } from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function read() {
+      try {
+        const response = await fetch('/api/todos');
+        if (!response.ok) throw new Error(`Error: ${response.status}`);
+
+        const todoRead = await response.json();
+        setTodos(todoRead);
+      } catch (err) {
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+
+    read();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const response = await fetch('/api/todos', {
+        method: 'POST',
+        body: JSON.stringify(newTodo),
+        headers: { 'Content-Type': 'application/json' },
+      });
+      if (!response.ok) throw new Error(`Error: ${response.status}`);
+      const addedTodo = await response.json();
+      const updateTodos = structuredClone(todos);
+      updateTodos.push(addedTodo);
+      setTodos(updateTodos);
+    } catch (err) {
+      setError(err);
```

Comment:
It's hard to get a nice UI with `setError` when there are errors adding or toggling. With setError, the whole page disappears and is replaced with the error message. Using an alert or a Toast usually works better.